src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -628,26 +594,61 @@ private void SplitHeaderFramesToOutput(int streamId, bool endOfHeaders, bool isF
         }
     }
 
-    private void FinishWritingHeadersUnsynchronized(int streamId)
+    private void FinishWritingHeadersUnsynchronized(int streamId, int payloadLength, HPackHeaderWriter.HeaderWriteResult writeResult, HttpResponseTrailers? trailers = null)
```

Comment:
I have to pass in the trailers here, so that I can re-initialize the _headersEnumerator (L625).
`Reset()` does not work on the `_headersEnumerator` because the underlying `Enumerator` of `HttpResponseTrailers`'s  does not 'reset' the `unknownEnumerator`. Then I recalled that 'Resetting' an enumerator isn't anyway a good idea and many BCL types throw.

And I have the impression if I move out the corresponding lines L619-L631 from here into the TRAILERS method or a new method, then we end up separating this or going back to what originally have.